c10/core/impl/cow/simulator.h

```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include <c10/macros/Macros.h>
+#include <c10/util/intrusive_ptr.h>
+
+#include <cstdint>
+#include <mutex>
+#include <type_traits>
+
+namespace c10::impl::cow {
+
+namespace detail {
+
+// An empty struct to use as the base class for a non-intrusive
+// SimulatorImpl.
+struct Empty {};
+
+} // namespace detail
+
+// Simulates a copy-on-write storage for a tensor.
+//
+// This simulator is used to identify situations where a copy-on-write
+// tensor would have been created and whether or not a logical copy
+// uses the storage after a different copy has modified it. So in
+// theory, it can identify reads/writes to copies after a write to a
+// different copy that is presently an alias.
```

Comment:
nit: want to know what a simulator semantically represents. In this case, it would be the "shadow storage" representing what the storage would look like after the CoW change

Reply:
reworded this, hopefully more clear.